### PR TITLE
libsleigh: Add helpers for finding spec files

### LINKED ISSUE #12: Add helper to locate SLA files
As per @pgoodman's comment at https://github.com/lifting-bits/sleigh/pull/9#issuecomment-948047563, we should look into a way of exposing a helper that finds SLA files in the SLEIGH installation that it belongs to. Ideally it should just look like this:
```c++
/*
* Returns a string containing the path to the .sla file for a given architecture.
* @param arch: The architecture for which we want to find a .sla file
* @return: An absolute path to the .sla file
* @throw std::invalid_argument: An unknown arch is provided.
*/
std::string FindSlaFile(const std::string& arch);
```

### PR DESCRIPTION
**Pull request by tetsuo-cpp**
Closes #12

### CODE REVIEW

**Comment by tetsuo-cpp** [comment]
I still need to tweak the installation to put all `.sla` and `.pspec` files under a few directories so that they can go in `gSearchPaths`. Let's get #21 in first as that influences it.

**Comment by tetsuo-cpp** [comment]
The Clang version on the Ubuntu workflow doesn't seem to support `std::filesystem` so I'll need to figure that out too.

**Comment by tetsuo-cpp** [comment]
Ok, so I ended up leaving the directory structure alone and leaning on the `std::filesystem` stuff to figure out where the files are.

I still need to figure out a nice way to keep CI running.

File: include/libsleigh.hh [hunk 1]
```diff
@@ -33,3 +33,63 @@
...
+
+static const char *gSearchPaths[] = {
+    // Derived from the installation
+    SLEIGH_SPEC_INSTALL_DIR "\0",
+    // Derived from the build
+    SLEIGH_SPEC_BUILD_DIR "\0",
+    // Common install locations
+    "/usr/local/share/sleigh", "/usr/share/sleigh", "/share/sleigh"};
+
+std::string FindSpecFileInSearchPath(std::string_view file_name,
```
Comment (tetsuo-cpp) [inline include/libsleigh.hh:47]: Since it's just a few helpers, I just put it straight under the includes in this mega-header. If you feel that this deserves its own TU, I can make a `.cxx` somewhere containing this stuff.

File: CMakeLists.txt [hunk 2]
```diff
@@ -190,6 +190,11 @@ target_compile_features(sleigh_settings INTERFACE
   cxx_std_17
 )
 
+target_compile_definitions(sleigh_settings INTERFACE
+  "SLEIGH_SPEC_INSTALL_DIR=\"${CMAKE_INSTALL_PREFIX}/share/sleigh/\""
+  "SLEIGH_SPEC_BUILD_DIR=\"${CMAKE_CURRENT_BINARY_DIR}/Ghidra\""
+)
+
```
Comment (ekilmer) [inline CMakeLists.txt:197]: This is a good start if the user has built and installed sleigh into the same paths, but I'd really like to support relocatable installations, like where you can build everything, tar it up, extract the tar to an arbitrary prefix and still be able to relatively easily find the spec files without modifications.

This would require the `sleigh-lift` executable to know where it is located (maybe [`cpplocate`](https://github.com/cginternals/cpplocate) or equivalent would help). If someone is building a new binary, then we should expose the location of the spec files through a CMake variable in the installed `sleighConfig.cmake` file.

The above doesn't need to be implemented in this PR, but it's something to keep in mind.

File: tools/sleigh-lift/src/main.cpp [hunk 3]
```diff
@@ -140,7 +140,7 @@ int main(int argc, char *argv[]) {
   }
   // Parse arguments
   const std::string action = argv[1];
-  const char *sla_file_path = argv[2];
+  const char *sla_file_name = argv[2];
```
Comment (ekilmer) [inline tools/sleigh-lift/src/main.cpp:143]: I think we should still let the user specify a path to search for compiled spec files, like before, but have it be optional. If that directory path is provided, search that path and error if not found.

**Review by ekilmer — CHANGES_REQUESTED** [review]

File: include/libsleigh.hh (hunk 1 above)
Comment (pgoodman) [reply include/libsleigh.hh:47]: I'd prefer you put the declaration of `FindSpecFile..` in a this header, but the implementation in a `.cpp`. You can also do something like:

header:
```
extern const std::vector<std::filesystem::path> kDefaultSlaSearchPaths;

std::filesystem::path FindSpecFile(
    std::string_view file_name,
    const std::vector<std::filesystem::path> &search_paths = kDefaultSlaSearchPaths);
```

Then define the default search paths in the `.cpp` file.

**Review by pgoodman — CHANGES_REQUESTED** [review]

File: tools/sleigh-lift/src/main.cpp (hunk 3 above)
Comment (tetsuo-cpp) [reply tools/sleigh-lift/src/main.cpp:143]: The start address is an optional positional argument right now and I'd rather not bring in GFlags. Also, `getopt` is a no-go since it'll make MSVC support harder.

I might try something dumber like allow a special `-f` argument at the beginning to allow the path to be specified.

File: CMakeLists.txt (hunk 2 above)
Comment (pgoodman) [reply CMakeLists.txt:198]: @tetsuo-cpp can you use a configure command instead of doing things as remill did with these? The macro way works (and is how it's done in remill, and I regret this). I'll share an example of configuration on Slack.

Comment (tetsuo-cpp) [reply CMakeLists.txt]: 2d632ec

I think this does the trick.

File: tools/sleigh-lift/src/main.cpp (hunk 3 above)
Comment (tetsuo-cpp) [reply tools/sleigh-lift/src/main.cpp]: 298ff76

File: CMakeLists.txt (hunk 2 above)
Comment (pgoodman) [reply CMakeLists.txt]: Can you configure a private header inside of your `support` directory, so that the main `Support.cpp` is part of the codebase and subject to IDE syntax coloring and stuff.

Similarly, can you make the search paths `std::filesystem::path`? Then `gSearchPaths` (renamed to `gDefaultSearchPaths`) can be exported in `Support.h` as:

```c++
extern std::vector<std::filesystem::path> gDefaultSearchPaths;

std::optional<std::filesystem::path>
FindSpecFile(std::string_view file_name,
             const std::vector<std::filesystem::path> &search_paths=gDefaultSearchPaths);
```

Then in Support.cpp:
```c++
const std::vector<std::filesystem::path> gDefaultSearchPaths{
    // Derived from the installation
    kSleighSpecInstallDir,
    // Derived from the build
    kSleightSpecBuildDir,
    // Common install locations
    "/usr/local/share/sleigh",
    "/usr/share/sleigh",
    "/share/sleigh"
};

std::optional<std::filesystem::path>
FindSpecFile(std::string_view file_name,
             const std::vector<std::filesystem::path> &search_paths) {
  // Try paths supplied by the caller first
  for (const auto &path : search_paths) {
    auto file_path = FindSpecFileInSearchPath(file_name, path);
    if (file_path) {
      return file_path;
    }
  }
  // Cannot find the spec file
  return {};
}
```

So support wouldn't search any default paths as a fallback because they would be in `search_paths` as a default-initialized parameter.

Comment (tetsuo-cpp) [reply CMakeLists.txt]: That doesn't seem very ergonomic to me. If you want to add to more paths to search, then you need to do something like this:
```c++
auto search_paths = sleigh::gDefaultSearchPaths;
search_paths.push_back(path1);
search_paths.push_back(path2);
const auto path = sleigh::FindSpecFile(file_name, search_paths);
```

Comment (pgoodman) [reply CMakeLists.txt]: My thinking was that if you knew where it was, then you wouldn't use the default paths. The other way of looking at it is that it seems surprising that it might go and search all over the place, even when you specify a particular list. Worst case, they don't need to copy and extend it; they can instead to do two calls, e.g.

```c++
auto maybe_path = sleigh::FindSpecFile(file_name, my_search_paths);
if (!maybe_path) {
  maybe_path = sleigh::FindSpecFile(file_name);
}
```

Comment (tetsuo-cpp) [reply CMakeLists.txt]: Makes sense, you're right.

952881a

Comment (tetsuo-cpp) [reply CMakeLists.txt]: Regarding this:

> Can you configure a private header inside of your support directory, so that the main Support.cpp is part of the codebase and subject to IDE syntax coloring and stuff.

What IDE are you using? I just tried with CLion and it seems to highlight `Support.cpp` correctly. And jumping to definition from the `sleigh-lift` example takes me to the right code.

Comment (pgoodman) [reply CMakeLists.txt]: Eclipse CDT, but where I don't let it touch compile_commands.json, and I use out-of-tree builds. Generally, if there are auto-generated files, I prefer them to be `#include`d so that I can introduce a new search path to my preferences, and have all the actual source all in one place.

File: README.md [hunk 4]
```diff
@@ -78,20 +78,20 @@ cmake --build build --target package
 An example program called `sleigh-lift` has been included to demonstrate how to use the SLEIGH API. It takes a hexadecimal string of bytes and either disassembles it or lifts it to p-code. The program can be invoked like so, where the `action` argument must be either `disassemble` or `pcode`:
 
 ```sh
-sleigh-lift [action] [sla_file] [bytes] [address:OPTIONAL]
+sleigh-lift [action] [sla_file] [bytes] [-a address] [-f sla_path]
```
Comment (ekilmer) [inline README.md:81]: `sla_path` is a bit ambiguous. Is it the path to the specific compiled `*.sla` file or is it to the root directory where we try to find `Ghidra/Processors/*/data/language`? The `-f` makes it seem like a "file" but then why is this here if we already specify `sla_file`? This should be either `sla_file_path` or `root_sla_dir_path` or something.

File: include/libsleigh.hh [hunk 5]
```diff
@@ -33,3 +34,5 @@
 #include "types.h"
 #include "xml.hh"
 #pragma GCC diagnostic pop
+
+#include "Support.h"
```
Comment (ekilmer) [inline include/libsleigh.hh:38]: This feels odd to include here because the support header is not actually a part of NSA's Ghidra/Sleigh. I feel like it would make more sense to keep the support header separate from this file.

Also, I think this causes issues if someone only links against `sleigh::sla`, since the `sleigh::support` target isn't a `PUBLIC` link in the CMake, so the `libsupport.a` will never be linked and will fail to compile.

File: support/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+add_library(support
+  ${sleigh_support_source_list}
+)
+
+add_library(sleigh::support ALIAS support)
+
+target_link_libraries(support PRIVATE
+  sleigh::sleigh_settings
+)
```
Comment (ekilmer) [inline support/CMakeLists.txt:28]: I forget the exact reason why the other libraries link the settings as `PUBLIC`, but I think we should be consistent and do it here too.

```suggestion
target_link_libraries(support PUBLIC
  sleigh::sleigh_settings
)
```

Side-note: the `sleigh_settings` target is really weird and I think we should remove the target in favor of CMake's [`set_target_properties`](https://cmake.org/cmake/help/latest/command/set_target_properties.html) where we can do pretty much the same thing without having to link to this weird settings pseudo-library and worry about correctly propagating the settings.

File: support/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+target_include_directories(support PRIVATE "${CMAKE_CURRENT_LIST_DIR}")
```
Comment (ekilmer) [inline support/CMakeLists.txt:29]: This should be conditional and only appear during the build interface methinks
```suggestion
target_include_directories(support PRIVATE $<BUILD_INTERFACE:"${CMAKE_CURRENT_LIST_DIR}>")
```

File: support/Support.cpp.in [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+} // namespace
+
+const std::vector<std::filesystem::path> gDefaultSearchPaths = {
+    // Derived from the installation
+    kSleighSpecInstallDir,
+    // Derived from the build
+    kSleighSpecBuildDir,
+    // Common install locations
+    "/usr/local/share/sleigh", "/usr/share/sleigh", "/share/sleigh"};
```
Comment (ekilmer) [inline support/Support.cpp.in:46]: these should change to point to the specfiles root directory
```suggestion
    "/usr/local/share/sleigh/specfiles", "/usr/share/sleigh/specfiles", "/share/sleigh/specfiles"};
```

**Review by ekilmer — CHANGES_REQUESTED** [review]

File: include/libsleigh.hh (hunk 5 above)
Comment (tetsuo-cpp) [reply include/libsleigh.hh:38]: Isn't that the case with not linking against `sleigh::decomp` too? It's just a convenience header so I figured it'd make sense to just include the helpers that you'd typically use alongside the library.

File: support/Support.cpp.in [hunk 9]
```diff
@@ -0,0 +1,62 @@
+#include "Support.h"
```
Comment (pgoodman) [inline support/Support.cpp.in]: Copyright block needed.

File: include/libsleigh.hh (hunk 5 above)
Comment (pgoodman) [reply include/libsleigh.hh:38]: I am in agreement with @tetsuo-cpp.

**Review by pgoodman — CHANGES_REQUESTED** [review]

File: README.md (hunk 4 above)
Comment (pgoodman) [reply README.md:81]: perhaps `sla_dir` if it's mean to be a directory?

Comment (tetsuo-cpp) [reply README.md]: Yeah it's the spot where we try to find `Ghidra/Processors/*/data/language`. I've renamed to `root_sla_dir` and changed the flag to `-p`.

9215bfb

File: include/libsleigh.hh (hunk 5 above)
Comment (ekilmer) [reply include/libsleigh.hh:38]: Fair enough 👍

File: support/Support.cpp [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+                         std::filesystem::path search_path) {
+  search_path.append("Ghidra").append("Processors");
+  // Check whether a SLEIGH installation exists at this path
+  if (!std::filesystem::is_directory(search_path)) {
+    return {};
+  }
+  // Each directory under Processors/ represents a family of architectures
+  //
+  // Spec files should reside under:
+  // <install_prefix>/Processors/<arch>/data/languages
```
Comment (ekilmer) [inline support/Support.cpp:28]: I think this should read more like 
```suggestion
  // <root_sla_dir>/Ghidra/Processors/<arch>/data/languages
```

**Review by ekilmer — APPROVED** [review]
One minor fix and LGTM!

Thank you!

File: support/Support.cpp (hunk 10 above)
Comment (ekilmer) [reply support/Support.cpp:28]: Not sure if you want to also change the line
```diff
-  // Each directory under Processors/ represents a family of architectures
+  // Each directory under Ghidra/Processors/ represents a family of architectures
```

**Review by pgoodman — APPROVED** [review]
